### DISCUSSION TITLE
What is the most efficient way to generate a large number of similarly designed Likert items?

### DISCUSSION THREAD

**Discussion by ctreffe** [ctreffe/alfred#95] [Alfred Q&A]
Especially when working with questionnaires containing a large number of items, users might wonder how to create these items in an efficient manner. Ideally, we want to use as few lines of code as possible and - as always - the solution should avoid duplicate code. In earlier versions of alfred (v2.x) there was a _LikertListElement()_ for this purpose, but there is no pendant in the new version.

How can this be achieved in __alfred3__?

**Comment by ctreffe** [comment]
In alfred2, the _LikertListElement()_ was a good solution to create questionnaires etc., but it was also known to cause problems with data saving and its display options were limited. Since one of the main reasons to code an experiment with Python is to benefit from the power of a programming language, we rely on coding solutions for this type of task in __alfred3__.

__Using a *for loop* presents a straightforward solution to this problem. The following minimal example will generate a number of neatly arranged Likert items with short and maintainable code.__

Let's start by creating a fresh alfred3 experiment. As we might want to randomize the order of our items later on, we also import _random_ from the Standard Library:

```Python
import alfred3 as al
import random

exp = al.Experiment()
```

As a first step in the creation of our questionnaire, we create a list of our item texts and their corresponding variable names. The list elements are created as tuples with the variable name in front:

```Python
questionnaire_items = [
    ("item_01", "This is the first item of the questionnaire"),
    ("item_02", "This is the second item of the questionnaire"),
    ("item_03", "This is the third item of the questionnaire"),
    ("item_04", "This is the fourth item of the questionnaire"),
    ("item_05", "This is the fifth item of the questionnaire"),
    ("item_06", "This is the sixth item of the questionnaire"),
    ("item_07", "This is the seventh item of the questionnaire"),
    ("item_08", "This is the eighth item of the questionnaire"),
    ("item_09", "This is the ninth item of the questionnaire"),
    ("item_10", "This is the tenth item of the questionnaire")
]

random.shuffle(questionnaire_items)  # Optional randomization of item order
```

The list could (and probably should) be moved to an extra content file, even more so when we create a large number of items. If we want to randomize the order of items, we need make sure that the last line of the above code is executed in each experiment session (e.g., by moving it into a Section's _on\_enter()_ hook).

It is time to create a page for our questionnaire:

```Python
questionnaire = al.Page(
    name="questionnaire",
    title="A Simple Questionnaire"
)
```

And now to the crucial part of the example. By using the for loop to iterate over our previously created list, we can avoid duplicate code and create as many Likert items as we wish. Within the loop, a _Text_ element and a _SingleChoiceBar_ element are created and arranged within a _Row_ element. The Row element gives us control over the horizontal arrangement of the item texts and the Likert items. In addition, _Row_ elements ensure a consistent layout for all questionnaire items on our page:

```Python
for item in questionnaire_items:
    # Text element to display the item text
    item_text = al.Text(item[1], width="full")

    # SingleChoiceBar for user input
    item_choice = al.SingleChoiceBar(
        1, 2, 3, 4, 5,
        name=item[0],
        leftlab="very bad",
        rightlab="very good"
    )

    # Important: We use a Row element to arrange text and input horizontally
    questionnaire += al.Row(
        item_text, item_choice,
        layout = (5,7)  # Adjust to your own liking
    )
```

The last thing we need to do is to add our questionnaire page to our experiment and add the lines of script code, which will allow us to start the experiment just by executing the _script.py_:

```Python
exp += questionnaire

if __name__ == "__main__":
    exp.run()
```

And that's it!  This example can be extended to any number of items. Using the _shuffle()_ method as shown above, we can also randomize the order of items for each participant without any side effects for the experiment dataset.

__A final word of advice:__  This solution iterates over a list of tuples instead of a dictionary to allow for an easy randomization of items. If randomization is not necessary, users might prefer iterating over a dictionary with the d.items() method:

```Python

questionnaire_items = {
    "item_01": "This is the first item of the questionnaire",
    "item_02": "This is the second item of the questionnaire",
    "item_03": "This is the third item of the questionnaire",
}

for key, value in questionnaire_items.items():
    ... # Take the for loop from above and use key instead of item[0] and value instead of item[1]

```

**Reply by jobrachem** [reply]
# Addition: Use a dictionary to future-proof your item data

@ctreffe showed a method that works perfectly well, using a list of *tuples*. We can extend this approach by using a list of *dictionaries* for the same purpose. This has two advantages:

1. The dictionary contains key-value pairs. That is a big advantage if you revisit your code after a few month or share it with collaborators (or online): It is much easier to understand, what the content of your list means.
2. Your can use Python's `**kwargs`` syntax to include the content elegantly in your element calls.

### Here's an example:

```python
import alfred3 as al

exp = al.Experiment()

my_items = [
    {"name": "item01", "toplab": "Instruction for item 1"},
    {"name": "item02", "toplab": "Instruction for item 2"}
]

@exp.member
class Demo(al.Page):
    def on_exp_access(self):
        for kwargs in my_items: 
            self += al.TextEntry(**kwargs)

if __name__ == "__main__":
    exp.run()
```

### How does the `**kwargs` thing work?

This is a shortcut that python offers us. *Inside* function calls or class instantiations, we can place double-asteriks in front of a dictionary. Python will then use the keys as argument names and the values as the argument values in our call. For example, the following three calls are fully equivalent:

```python
import alfred3 as al

# option 1, using double-asteriks
my_args = {"name": "demo01", "toplab": "My lab"}
al.Text(**my_args)

# option 2, manual access from dictionary
my_args = {"name": "demo01", "toplab": "My lab"}
al.Text(name=my_args["name"], toplab=my_args["toplab"])

# option 3, using the ordinary way
al.Text(name="demo01", toplab="My lab")
```

Now, option 1 is much better than option 2, because we have to do a lot less typing. Option 3 is of course nice in any case in which you define an individual element.

You can also combine options 1 and 3:

```python
import alfred3 as al

# combination of options 1 and 3
my_args = {"name": "demo01", "toplab": "My lab"}
al.Text(force_input=True, **my_args)
```

### Why do you use `**kwargs` in the first example, and then `**my_args` in the others?

Because the important part is the double-asterik (`**`). What comes after the double-asterik is just a reference to the specific dictionary that you want to "unpack". In the first example, I  use "kwargs", because it is the standard shorthand for "keyword arguments" in Python.

### Can I place the list of item arguments in an external file?

Yes, you can. For example, we could place our `my_items` in a `content.py` in a subdirectory of our experiment directory.

```python
# files/content.py
my_items = [
    {"name": "item01", "toplab": "Instruction for item 1"},
    {"name": "item02", "toplab": "Instruction for item 2"}
]
```

Next, we "smuggle" this file into our `script.py` and access the content:

```python
import alfred3 as al
from thesmuggler import smuggle  # import smuggle

content = smuggle("files/content.py").  # load external file

exp = al.Experiment()


@exp.member
class Demo(al.Page):
    def on_exp_access(self):
        for kwargs in content.my_items:  # change access to list object
            self += al.TextEntry(**kwargs)

if __name__ == "__main__":
    exp.run()
```
